File: app-tracker-automation/pages/aditya_birla_login_page.py
"""
Aditya Birla Sun Life Insurance Login Page
Enhanced login page for UAT environment
"""

from playwright.sync_api import Page, expect
from pages.base_page import BasePage
from locators.aditya_birla_locators import AdityaBirlaLocators
from utils.logger import Logger
from utils.waits import Waits


class AdityaBirlaLoginPage(BasePage):
    """Aditya Birla Sun Life Insurance login page"""
    
    def __init__(self, page: Page):
        super().__init__(page)
        self.locators = AdityaBirlaLocators()
        self.waits = Waits(page)
    
    def load(self):
        """Load Aditya Birla UAT login page"""
        self.logger.info("Loading Aditya Birla UAT login page")
        self.navigate_to(self.locators.login_page["url"])
        self.waits.wait_for_page_load()
    
    def is_login_page_displayed(self) -> bool:
        """Check if login page is displayed"""
        try:
            # Check URL
            current_url = self.page.url
            if "leapuat.adityabirlasunlifeinsurance.com" not in current_url:
                return False
            
            # Check key elements
            username_visible = self.is_element_visible(self.locators.login_page["username_input"])
            password_visible = self.is_element_visible(self.locators.login_page["password_input"])
            login_button_visible = self.is_element_visible(self.locators.login_page["login_button"])
            
            return username_visible and password_visible and login_button_visible
            
        except Exception as e:
            self.logger.error(f"Error checking login page display: {str(e)}")
            return False
    
    def enter_credentials(self, username: str, password: str):
        """Enter login credentials"""
        try:
            self.logger.info(f"Entering credentials for user: {username}")
            
            # Find and fill username
            username_input = self.page.locator(self.locators.login_page["username_input"]).first
            self.waits.wait_for_element_visible(self.locators.login_page["username_input"])
            username_input.fill(username)
            
            # Find and fill password
            password_input = self.page.locator(self.locators.login_page["password_input"]).first
            self.waits.wait_for_element_visible(self.locators.login_page["password_input"])
            password_input.fill(password)
            
            # Verify password masking
            password_type = password_input.get_attribute("type")
            if password_type != "password":
                self.logger.warning("Password field is not properly masked")
            
            self.logger.info("Credentials entered successfully")
            
        except Exception as e:
            self.logger.error(f"Failed to enter credentials: {str(e)}")
            raise
    
    def click_login_button(self):
        """Click the login button"""
        try:
            self.logger.info("Clicking login button")
            
            login_button = self.page.locator(self.locators.login_page["login_button"]).first
            self.waits.wait_for_element_enabled(self.locators.login_page["login_button"])
            login_button.click()
            
            self.logger.info("Login button clicked")
            
        except Exception as e:
            self.logger.error(f"Failed to click login button: {str(e)}")
            raise
    
    def login(self, username: str, password: str):
        """Perform complete login flow"""
        try:
            self.logger.info(f"Starting login flow for user: {username}")
            
            # Load login page
            self.load()
            
            # Verify page is displayed
            if not self.is_login_page_displayed():
                raise Exception("Login page is not properly displayed")
            
            # Enter credentials
            self.enter_credentials(username, password)
            
            # Click login
            self.click_login_button()
            
            # Wait for redirect
            self.page.wait_for_timeout(3000)
            
            # Check if login was successful (redirected away from login page)
            current_url = self.page.url
            if "#/login" in current_url:
                # Check for error message
                error_visible = self.is_element_visible(self.locators.login_page["error_message"])
                if error_visible:
                    error_text = self.get_error_message()
                    self.logger.error(f"Login failed with error: {error_text}")
                    return False
                else:
                    self.logger.error("Login failed - still on login page")
                    return False
            else:
                self.logger.info("Login successful - redirected from login page")
                return True
                
        except Exception as e:
            self.logger.error(f"Login flow failed: {str(e)}")
            return False
    
    def get_error_message(self) -> str:
        """Get login error message"""
        try:
            error_element = self.page.locator(self.locators.login_page["error_message"]).first
            if error_element.is_visible():
                return error_element.text_content()
            return ""
        except:
            return ""
    
    def verify_uat_badge(self) -> bool:
        """Verify UAT badge is visible"""
        return self.is_element_visible(self.locators.login_page["uat_badge"])
    
    def verify_branding(self) -> bool:
        """Verify Aditya Birla branding is visible"""
        return self.is_element_visible(self.locators.login_page["branding_header"])
    
    def wait_for_page_load(self, timeout: int = 10000):
        """Wait for login page to fully load"""
        try:
            # Wait for URL to contain login
            self.waits.wait_for_url_contains("#/login", timeout)
            
            # Wait for key elements to be visible
            self.waits.wait_for_element_visible(self.locators.login_page["username_input"], timeout)
            self.waits.wait_for_element_visible(self.locators.login_page["password_input"], timeout)
            self.waits.wait_for_element_visible(self.locators.login_page["login_button"], timeout)
            
            self.logger.info("Login page loaded successfully")
            
        except Exception as e:
            self.logger.error(f"Login page load timeout: {str(e)}")
            raise
    
    def clear_credentials(self):
        """Clear username and password fields"""
        try:
            username_input = self.page.locator(self.locators.login_page["username_input"]).first
            password_input = self.page.locator(self.locators.login_page["password_input"]).first
            
            username_input.fill("")
            password_input.fill("")
            
            self.logger.info("Credentials cleared")
            
        except Exception as e:
            self.logger.error(f"Failed to clear credentials: {str(e)}")
    
    def is_username_focused(self) -> bool:
        """Check if username field has focus"""
        try:
            username_input = self.page.locator(self.locators.login_page["username_input"]).first
            return username_input.is_focused()
        except:
            return False
    
    def is_password_focused(self) -> bool:
        """Check if password field has focus"""
        try:
            password_input = self.page.locator(self.locators.login_page["password_input"]).first
            return password_input.is_focused()
        except:
            return False
    
    def get_page_title(self) -> str:
        """Get page title"""
        return self.page.title()
    
    def take_screenshot(self, name: str = "login_page"):
        """Take screenshot of login page"""
        timestamp = self.page.evaluate("new Date().toISOString().replace(/[:.]/g, '-')")
        filename = f"{name}_{timestamp}.png"
        self.page.screenshot(path=f"screenshots/{filename}", full_page=True)
        self.logger.info(f"Screenshot saved: {filename}")

File: ts-framework/pages/aditya_birla_login_page.ts
import { Page } from '@playwright/test';
import { BasePage } from './base_page';
import { AdityaBirlaLocators } from '../locators/aditya_birla_locators';

export class AdityaBirlaLoginPage extends BasePage {
  private locators: AdityaBirlaLocators;

  constructor(page: Page) {
    super(page);
    this.locators = new AdityaBirlaLocators(page);
  }

  async load(): Promise<void> {
    this.logger.info('Loading Aditya Birla UAT login page');
    await this.navigateTo(this.locators.loginPageUrl);
    await this.waits.waitForPageLoad();

    // Wait for network to settle after page load
    await this.page.waitForLoadState('networkidle', { timeout: 10000 });

    // Wait for any loading overlays to disappear
    await this.waitForLoadingOverlayToDisappear();
  }

  async isLoginPageDisplayed(): Promise<boolean> {
    try {
      // Check URL
      const currentUrl = this.page.url();
      if (!currentUrl.includes('leapuat.adityabirlasunlifeinsurance.com')) {
        return false;
      }

      // Check key elements
      const usernameVisible = await this.locators.usernameInput.isVisible();
      const passwordVisible = await this.locators.passwordInput.isVisible();
      const loginButtonVisible = await this.locators.loginButton.isVisible();

      return usernameVisible && passwordVisible && loginButtonVisible;

    } catch (e) {
      this.logger.error(`Error checking login page display: ${String(e)}`);
      return false;
    }
  }

  async enterCredentials(username: string, password: string): Promise<void> {
    try {
      this.logger.info(`Entering credentials for user: ${username}`);

      const usernameInput = this.locators.usernameInput.first();
      await this.waits.waitForElementVisible(this.locators.usernameInput);

      // Wait for element to be fully actionable
      await usernameInput.waitFor({ state: 'attached', timeout: 5000 });
      await usernameInput.waitFor({ state: 'visible', timeout: 5000 });

      await usernameInput.fill(username);

      // Wait for network to settle after username input
      await this.page.waitForLoadState('networkidle', { timeout: 3000 });

      const passwordInput = this.locators.passwordInput.first();
      await this.waits.waitForElementVisible(this.locators.passwordInput);

      // Wait for element to be fully actionable
      await passwordInput.waitFor({ state: 'attached', timeout: 5000 });
      await passwordInput.waitFor({ state: 'visible', timeout: 5000 });

      await passwordInput.fill(password);

      // Verify password masking
      const passwordType = await passwordInput.getAttribute('type');
      if (passwordType !== 'password') {
        this.logger.warn('Password field is not properly masked');
      }

      // Wait for network to settle after password input
      await this.page.waitForLoadState('networkidle', { timeout: 3000 });

      this.logger.info('Credentials entered successfully');

    } catch (e) {
      this.logger.error(`Failed to enter credentials: ${String(e)}`);
      throw e;
    }
  }

  async clickLoginButton(): Promise<void> {
    try {
      this.logger.info('Clicking login button');

      const loginButton = this.locators.loginButton.first();
      
      // NATIVE PLAYWRIGHT WAITS: Replaces custom this.waits method to fix Type mismatch
      await loginButton.waitFor({ state: 'attached', timeout: 5000 });
      await loginButton.waitFor({ state: 'visible', timeout: 5000 });

      // Explicitly check if it is enabled before clicking
      if (!(await loginButton.isEnabled())) {
         throw new Error("Login button is present but disabled.");
      }

      // Wait for any loading overlays to disappear
      await this.waitForLoadingOverlayToDisappear();

      await loginButton.click();

      this.logger.info('Login button clicked');

    } catch (e) {
      this.logger.error(`Failed to click login button: ${String(e)}`);
      throw e;
    }
  }

  async login(username: string, password: string): Promise<boolean> {
    try {
      this.logger.info(`Starting login flow for user: ${username}`);

      // Load login page
      await this.load();

      // Verify page is displayed
      if (!(await this.isLoginPageDisplayed())) {
        throw new Error('Login page is not properly displayed');
      }

      // Enter credentials
      await this.enterCredentials(username, password);

      // Click login
      await this.clickLoginButton();

      // Wait for redirect
      await this.page.waitForTimeout(3000);

      // Check if login was successful (redirected away from login page)
      const currentUrl = this.page.url();
      if (currentUrl.includes('#/login')) {
        // Check for error message
        const errorVisible = await this.locators.errorMessage.isVisible();
        if (errorVisible) {
          const errorText = await this.getErrorMessage();
          this.logger.error(`Login failed with error: ${errorText}`);
          return false;
        } else {
          this.logger.error('Login failed - still on login page');
          return false;
        }
      } else {
        this.logger.info('Login successful - redirected from login page');
        return true;
      }

    } catch (e) {
      this.logger.error(`Login flow failed: ${String(e)}`);
      return false;
    }
  }

  async getErrorMessage(): Promise<string> {
    try {
      const errorElement = this.locators.errorMessage.first();
      if (await errorElement.isVisible()) {
        const text = await errorElement.textContent();
        return text ? text.trim() : '';
      }
      return '';
    } catch {
      return '';
    }
  }

  async verifyUatBadge(): Promise<boolean> {
    return await this.locators.uatBadge.isVisible();
  }

  async verifyBranding(): Promise<boolean> {
    return await this.locators.brandingHeader.isVisible();
  }

  async waitForPageLoad(timeout: number = 10000): Promise<void> {
    try {
      // Wait for URL to contain login
      await this.waits.waitForUrlContains('#/login', timeout);

      // Wait for key elements to be visible
      await this.waits.waitForElementVisible(this.locators.usernameInput, timeout);
      await this.waits.waitForElementVisible(this.locators.passwordInput, timeout);
      await this.waits.waitForElementVisible(this.locators.loginButton, timeout);

      this.logger.info('Login page loaded successfully');

    } catch (e) {
      this.logger.error(`Login page load timeout: ${String(e)}`);
      throw e;
    }
  }

  async clearCredentials(): Promise<void> {
    try {
      const usernameInput = this.locators.usernameInput.first();
      const passwordInput = this.locators.passwordInput.first();

      await usernameInput.fill('');
      await passwordInput.fill('');

      this.logger.info('Credentials cleared');

    } catch (e) {
      this.logger.error(`Failed to clear credentials: ${String(e)}`);
      throw e;
    }
  }

  async isUsernameFocused(): Promise<boolean> {
    try {
      const usernameInput = this.locators.usernameInput.first();
      // FIX: Ask the browser if this element is the currently active/focused element
      return await usernameInput.evaluate((node) => document.activeElement === node);
    } catch {
      return false;
    }
  }

  async isPasswordFocused(): Promise<boolean> {
    try {
      const passwordInput = this.locators.passwordInput.first();
      // FIX: Ask the browser if this element is the currently active/focused element
      return await passwordInput.evaluate((node) => document.activeElement === node);
    } catch {
      return false;
    }
  }

  async getPageTitle(): Promise<string> {
    return await this.page.title();
  }

  private async waitForLoadingOverlayToDisappear(timeout: number = 5000): Promise<void> {
    try {
      // Common loading overlay selectors
      const loadingSelectors = [
        '.loading-overlay',
        '.spinner',
        '.progress-bar',
        "[class*='loading']",
        "[class*='spinner']",
        "[class*='overlay']",
        '.MuiCircularProgress-root',
        '.MuiBackdrop-root'
      ];

      for (const selector of loadingSelectors) {
        try {
          const loadingElement = this.page.locator(selector).first();
          if (await loadingElement.isVisible({ timeout: 1000 })) {
            this.logger.info(`Waiting for loading element to disappear: ${selector}`);
            await loadingElement.waitFor({ state: 'hidden', timeout });
            this.logger.info(`Loading element disappeared: ${selector}`);
          }
        } catch {
          continue;
        }
      }
    } catch (e) {
      this.logger.error(`Error waiting for loading overlay: ${String(e)}`);
    }
  }
}